### PR TITLE
DNSSEC17 test case implementation

### PR DESCRIPTION
**Pull request by vlevigneron**
The code is not ready to be merged
Some unit test are still missing :
- DS17_CDNSKEY_MATCHES_NO_DNSKEY
- DS17_CDNSKEY_SIGNED_BY_UNKNOWN_DNSKEY
- DS17_DELETE_CDNSKEY
- DS17_MIXED_DELETE_CDNSKEY

----------------------------------------------------------------------------------------------------------------------
Updated 2021-05-18 by @matsduf:
## How to test this PR

Read the test case specification for DNSSEC17 and verify that the implementation correctly reports issues and non-issues on selected zones to cover as many message tags as possible.

### CODE REVIEW

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 1]
```diff
@@ -716,6 +732,56 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    DS17_DNSKEY_NOT_SIGNED_BY_CDNSKEY => sub {
+        __x    # DS17_DNSKEY_NOT_SIGNED_BY_CDNSKEY
+          'The DNSKEY RRset is not signed by the DNSKEY that the CDNSKEY record with tag '
+          . '{keytag} points to. Fetched from the nameservers with IP addresses '
+          . '"{ns_ip_list}".',
+          @_;
+    },
+    DS17_MIXED_DELETE_CDNSKEY => sub {
+        __x    # DS17_MIXED_DELETE_CDNSKEY
+          'The CDS RRset is a mixture between "delete" record and other records. '
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:781]: "CDS" => "CDNSKEY"

File: share/profile_example.json [hunk 2]
```diff
@@ -239,6 +239,14 @@
...
             "DS16_DNSKEY_NOT_SIGNED_BY_CDS" : "WARNING",
             "DS16_MIXED_DELETE_CDS" : "ERROR",
+            "DS17_CDNSKEY_INVALID_RRSIG" : "ERROR",
+            "DS17_CDNSKEY_MATCHES_NO_DNSKEY" : "WARNING",
+            "DS17_CDNSKEY_SIGNED_BY_UNKNOWN_DNSKEY" : "ERROR",
+            "DS17_CDNSKEY_UNSIGNED" : "ERROR",
+            "DS17_CDNSKEY_WITHOUT_DNSKEY" : "ERROR",
+            "DS17_DELETE_CDNSKEY" : "INFO",
+            "DS17_DNSKEY_NOT_SIGNED_BY_CDNSKEY" : "WARNING",
+            "DS17_MIXED_DELETE_CDNSKEY" : "ERROR",
```
Comment (matsduf) [inline share/profile_example.json:249]: Below, after line 491, there should be a line "dnssec17", shouldn't there?

**Review by matsduf — CHANGES_REQUESTED** [review]

File: t/Test-dnssec.t [hunk 3]
```diff
@@ -400,6 +400,33 @@ $zone = Zonemaster::Engine->zone( 'dnssec16-cds-matches-no-dnskey.zft-root.rd.ni
...
+
+$zone = Zonemaster::Engine->zone( 'dnssec16-cds-cdnskey-invalid-rrsig-01.zft-root.rd.nic.fr' );
+zone_gives( 'dnssec17', $zone, [q{DS17_CDNSKEY_INVALID_RRSIG}]);
+zone_gives_not('dnssec17', $zone, [qw{DS17_CDNSKEY_MATCHES_NO_DNSKEY DS17_CDNSKEY_SIGNED_BY_UNKNOWN_DNSKEY DS17_CDNSKEY_UNSIGNED DS17_CDNSKEY_WITHOUT_DNSKEY DS17_DELETE_CDNSKEY DS17_DNSKEY_NOT_SIGNED_BY_CDNSKEY DS17_MIXED_DELETE_CDNSKEY}] );
+
+$zone = Zonemaster::Engine->zone( 'dnssec16-cds-cdnskey-invalid-rrsig-02.zft-root.rd.nic.fr' );
+zone_gives( 'dnssec17', $zone, [q{DS17_CDNSKEY_INVALID_RRSIG}]);
+zone_gives_not('dnssec17', $zone, [qw{DS17_CDNSKEY_MATCHES_NO_DNSKEY DS17_CDNSKEY_SIGNED_BY_UNKNOWN_DNSKEY DS17_CDNSKEY_UNSIGNED DS17_CDNSKEY_WITHOUT_DNSKEY DS17_DELETE_CDNSKEY DS17_DNSKEY_NOT_SIGNED_BY_CDNSKEY DS17_MIXED_DELETE_CDNSKEY}] );
+
+$zone = Zonemaster::Engine->zone( 'dnssec17-cdnskey-unsigned.zft-root.rd.nic.fr' );
```
Comment (matsduf) [inline t/Test-dnssec.t:422]: Something is probably missing in delegation:
```
  20.08 WARNING   DNSSEC01       Nameserver vps430862.ovh.net/145.239.76.202 responded with an unexpected rcode (NXDOMAIN) on a DS query for zone dnssec17-cdnskey-unsigned.zft-root.rd.nic.fr.
  20.08 WARNING   DNSSEC02       Nameserver vps430862.ovh.net/145.239.76.202 responded with an unexpected rcode (NXDOMAIN) on a DS query for zone dnssec17-cdnskey-unsigned.zft-root.rd.nic.fr.
```

File: share/profile_example.json (hunk 2 above)
Comment (vlevigneron) [reply share/profile_example.json:249]: Yes. Fixed.

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 1 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:781]: Fixed

File: t/Test-dnssec.t (hunk 3 above)
Comment (vlevigneron) [reply t/Test-dnssec.t:422]: Fixed

**Comment by matsduf** [comment]
So far it looks good. Will the remaining unit tests be created, or should that be postponed? If postponed, an issue should be created stating that.

**Comment by vlevigneron** [comment]
@matsduf I will not have time this week for the missing unit tests. It will be better to merge without it and create an issue for that.

**Comment by vlevigneron** [comment]
Issue created https://github.com/zonemaster/zonemaster-engine/issues/907

**Review by matsduf — APPROVED** [review]

**Comment by matsduf** [comment]
Resolves #898

**Comment by matsduf** [comment]
## Release testing for v2021.1
Works as expected.

**Pull request merged**